presto-spi/src/main/java/com/facebook/presto/spi/block/BlockLeases.java

```diff
@@ -0,0 +1,139 @@
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array), () -> arrayAllocator.returnArray(nulls));
+    }
+
+    public static BlockLease newLease(short[] array, Block block, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array));
+    }
+
+    public static BlockLease newLease(short[] array, boolean[] nulls, Block block, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array), () -> arrayAllocator.returnArray(nulls));
+    }
+
+    public static BlockLease newLease(int[] array, Block block, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array));
+    }
+
+    public static BlockLease newLease(int[] array, boolean[] nulls, Block block, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array), () -> arrayAllocator.returnArray(nulls));
+    }
+
+    public static BlockLease newLease(long[] array, Block block, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array));
+    }
+
+    public static BlockLease newLease(long[] array, boolean[] nulls, Block block, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, () -> arrayAllocator.returnArray(array), () -> arrayAllocator.returnArray(nulls));
+    }
+
+    public static BlockLease newLease(Block block)
+    {
+        return new ClosingBlockLease(block);
+    }
+
+    public static BlockLease newLease(Block block, BlockLease blockLease)
+    {
+        return new ClosingBlockLease(block, blockLease::close);
+    }
+
+    public static BlockLease newLease(Block block, BlockLease... blockLeases)
+    {
+        Closer[] closers = Arrays.stream(blockLeases)
+                .map(i -> (Closer) i::close)
+                .toArray(Closer[]::new);
+        return new ClosingBlockLease(block, closers);
+    }
+
+    public static BlockLease newLease(Block block, BlockLease blockLease, int[] intArray, ArrayAllocator arrayAllocator)
+    {
+        return new ClosingBlockLease(block, blockLease::close, () -> arrayAllocator.returnArray(intArray));
+    }
+
+    private static final class ClosingBlockLease
+            implements BlockLease
+    {
+        private final Block block;
+        private final List<Closer> closers;
+
+        public ClosingBlockLease(Block block, Closer... closers)
+        {
+            this.block = block;
+            this.closers = unmodifiableList(asList(closers));
+        }
+
+        @Override
+        public Block get()
+        {
+            return block;
+        }
+
+        @Override
+        public void close()
+        {
+            closeAll(closers);
+        }
+    }
+
+    interface Closer
```

Comment:
AutoClosable throws an Exception, but we can override it to not throw an Exception.  I would like the intended semantic to be no exception, otherwise we'd want the exception handling to be propagated outside of BlockLease and it would make its usage more complicated.  It's only intended to request/return resources that may exist in-memory and so exception handling shouldn't enter the picture.